### ISSUE TITLE
Subflows - What does that measure?

### ISSUE THREAD

**Issue by swight-prc** [ahlashkari/CICFlowMeter#110]
I'm finding subflows to be confusing.
` BasicFlow.java beginning line 338`
```java
	private long sfLastPacketTS=-1;
	private int sfCount=0;
	private long sfAcHelper=-1;

	void detectUpdateSubflows( BasicPacketInfo packet ){
		if(sfLastPacketTS == -1){
			sfLastPacketTS = packet.getTimeStamp();
			sfAcHelper   = packet.getTimeStamp();
		}
		//System.out.print(" - "+(packet.timeStamp - sfLastPacketTS));
		if( (packet.getTimeStamp() - (sfLastPacketTS)/(double)1000000)   > 1.0 ){   //  What is this line doing?
			sfCount ++ ;
			long lastSFduration = packet.getTimeStamp() - sfAcHelper;
			updateActiveIdleTime(packet.getTimeStamp() - sfLastPacketTS, 5000000L);
			sfAcHelper = packet.getTimeStamp();
		}

		sfLastPacketTS = packet.getTimeStamp() ;
	}
```

Line 348 specifically:
```java
		if( (packet.getTimeStamp() - (sfLastPacketTS)/(double)1000000)   > 1.0 ){
```

If I'm reading this correctly, this will land for EVERY packet (even the first one), because you're subtracting `sfLastPacketTS` (converted to s) from  `packet.getTimeStamp()` (in µs), then checking to see if that's >1.

Was the goal to check whether 1s had passed between packets? That seems reasonable, and if so that could be fixed with careful application of parentheses.

I don't have a great understanding of network traffic to begin with, so I'm still trying to learn what the purpose of highlighting these subflows is. I came across this, and I thought I'd bring it up.

**Comment by swight-prc** [comment]
Since we're at it, why is this method calling `updateActiveIdleTime` at all, and when it does, why is it calling with a timedelta instead of a timestamp? Might explain why none of my flows have any active time...

**Comment by swight-prc** [comment]
Apparently based on an old commit; this has already been addressed.

**Comment by fisher85** [comment]
Tell me please, why did you decide to close the issue? I'm looking at the [BasicFlow.java](https://github.com/ahlashkari/CICFlowMeter/blob/c0e1039ebc2384fd059d1329cdbe23320bf4cdf8/src/main/java/cic/cs/unb/ca/jnetpcap/BasicFlow.java) code in the last commit and I can't figure out if the parentheses are correctly placed: `if( (packet.getTimeStamp() - (sfLastPacketTS)/(double)1000000)   > 1.0 ){`.

**Comment by swight-prc** [comment]
Let's look:
```java
if( (packet.getTimeStamp() - (sfLastPacketTS)/(double)1000000) > 1.0 ){
  1 2                   33   3              3 3      3       2       1
```
1's are the if statement; 2's are the entire math problem, and 3's are the innermost sets of parens.

So yeah, it looks like you're right. We need another set of parens. I saw that another issue had been fixed, and I didn't look close enough to see that this one was still problematic.

**Comment by ltkhang** [comment]
I'm converting this tool into python (a light version for my project here https://github.com/ltkhang/sdn-ids-ddos-defense) but the logical algo behind this tool has a lot of problems. Besides, they even have not declared or explained the whole process of this tool, I was finding the technical reports but there was no of them out there.

**Comment by mikonnikova** [comment]
This problem was fixed in commit 5df1a62 but without any notification.

**Comment by swight-prc** [comment]
> This problem was fixed in commit [5df1a62](https://github.com/ahlashkari/CICFlowMeter/commit/5df1a62e9cf16380c79c87779067f3c630ca8578) but without any notification.

Cool. Yeah, that does look like it fixes the issue.

Closing.
